src/Cache/FileSystemCacheItemPool.php

```diff
@@ -0,0 +1,212 @@
+<?php
+/**
+ * Copyright 2024 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+namespace Google\Auth\Cache;
+
+use ErrorException;
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Cache\CacheItemInterface;
+
+class FileSystemCacheItemPool implements CacheItemPoolInterface
+{
+    private string $cachePath = 'cache/';
+    private array $deferredPool = [];
+
+    public function __construct($options = [])
+    {
+        if (array_key_exists('path', $options)) {
+            $this->cachePath = $options['path'];
+        }
+
+        if (is_dir($this->cachePath)) {
+            return true;
+        }
+
+        if (!mkdir($this->cachePath)) {
+            throw new ErrorException("Cache folde couldn't be created");
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItem(string $key): CacheItemInterface
+    {
+        if (!$this->isValidKey($key)) {
+            throw new InvalidArgumentException("The key '$key' is not valid. The key should follow the pattern |^[a-zA-Z0-9_\.! ]+$|");
+        }
+
+        if (!$this->hasItem($key)) {
+            return new TypedItem($key);
+        }
+
+        // Could this cause any issues? Check for result first maybe?
+        return unserialize(file_get_contents($this->cacheFilePath($key)));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItems(array $keys = []): iterable
+    {
+        $result = [];
+
+        foreach ($keys as $key) {
+            if ($this->isValidKey($key)){
+                throw new InvalidArgumentException("The key '$key' is not valid. The key should follow the pattern |^[a-zA-Z0-9_\.! ]+$|");
+            }
+
+            $result[$key] = $this->getItem($key);
+        }
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasItem(string $key): bool
+    {
+        $itemPath = $this->cacheFilePath($key);
+
+        if (!file_exists($itemPath)) {
+            return false;
+        }
+
+        $serializedItem = file_get_contents($this->cacheFilePath($key));
```

Comment:
We can reuse the calculated path.

```suggestion
        $serializedItem = file_get_contents($itemPath);
```